### DISCUSSION TITLE
Version 1.0.0... at last!

### DISCUSSION THREAD

**Discussion by ZZ-Cat** [ZZ-Cat/CRSFforArduino#86] [Announcements]
## Overview

Wow, what a journey it has been!
The first `MAJOR` cut of CRSF for Arduino is finally here.  

I can't say this enough: A _huge_ thank you to you all for your contributions, hardware testing, Issues, Pull Requests, your patience with me, and your wonderful ideas in my Discussions tab.  
You all have helped make Version 1.0.0 of CRSF for Arduino a reality.

Thank you _very much_ from the bottom of my heart.  
This project has been some two years in the making (both offline and online), through various iterations and attempts.  
A few bumps along the way, plus a point where I simply wanted to give up, delete this repo and call it quits.  
But, I found it in me to keep going, despite the setbacks... and gods, that has finally paid off.  

Some say firmware and software development is a thank-less job.  
I beg to differ. My own experience, I have had nothing but overwhelming support from my user-base.  
You all have been what's kept me going through the times where I had reached my absolute lowest with CRSF for Arduino.  
You folks are amazing.  

So, here's to you all, my gift to you... CRSF for Arduino, Version 1.0.0.  
Enjoy~ 🖤 

### Features

* feat(telemetry): Add battery sensor telemetry by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/45
* feat(library): Add full telemetry support by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/47
* feat(link statistics): Add Link Statistics API by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/78
* feat(serial receiver interface): Add custom Flight Mode strings as telemetry by @berickson in https://github.com/ZZ-Cat/CRSFforArduino/pull/83

### Hardware targets

* feat(devboard): Add support for Teensy 3.x and 4.x by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/37
* feat(devboard): Add support for ESP32 by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/39
* feat(devboard): Add support for RP2040 by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/72
* feat(devboard) Add support for STM32 F103, F4, and H7 targets by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/81

### Fixes

* fix(hal): Inherit from Stream instead of HardwareSerial across all supported targets by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/44
* fix(platformio): Remove `$PROJECT_DIR` from all relevant paths. by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/64
* fix(compatibility table): Add missing Arduino Nano 33 IoT compatibility by @Speshl in https://github.com/ZZ-Cat/CRSFforArduino/pull/57
* fix(platformio): CRSF for Arduino not recognised as a dependency by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/77

### Re-factors and re-writes

* refactor(rc channels): Improve readability of RC Channels in Serial Plotter by @Kristabel28 in https://github.com/ZZ-Cat/CRSFforArduino/pull/40
* refactor(compatibility table): Permissive Incompatibility by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/68
* refactor(hal): Deprecate `DevBoards` class and bring `HardwareSerial` class up to the Sketch Layer by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/71
* refactor(rc channels): Use event handler for fail-safe and RC channels by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/79
* refactor(library): Flight Modes API and Custom Flight Modes integration, plus code and comments clean-up by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/84

### Chores

* chore(library): Restructure entire repository by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/46

### Miscellaneous

* build(ci): Trigger CI checks on pull requests by @ZZ-Cat in https://github.com/ZZ-Cat/CRSFforArduino/pull/41

## Kia ora and haere mai to our new contributors

* @Kristabel28 made their first contribution in https://github.com/ZZ-Cat/CRSFforArduino/pull/40
* @Speshl made their first contribution in https://github.com/ZZ-Cat/CRSFforArduino/pull/57
* @berickson made their first contribution in https://github.com/ZZ-Cat/CRSFforArduino/pull/83

**Full Changelog**: https://github.com/ZZ-Cat/CRSFforArduino/compare/v0.4.0...v1.0.0

## Semantic Versioning

All releases of CRSF for Arduino follow the [Semantic Versioning Guidelines.](https://semver.org/)
This means "Major.Minor.Patch".
Major = New features that are incompatible with previous iterations.
Minor = New features &/or hardware support that retains compatibility with the current Major release.
Patch = Bug fixes.

This is Major Version 1.x.x and is considered stable.  
While every effort has gone in to ensure the latest release is bug-free, there _are_ no guarantees with firmware development.  
Your feedback is always welcome. If you spot any bugs, don't hesitate to open an [issue.](https://github.com/ZZ-Cat/CRSFforArduino/issues)
For general comments & questions, please use the [discussions](https://github.com/ZZ-Cat/CRSFforArduino/discussions) section.  
New contributors are always welcome. Please read through my [Contribution Guidelines](https://github.com/ZZ-Cat/CRSFforArduino/blob/Main-Trunk/.github/CONTRIBUTING.md) _before_ you go Gung Ho with coding.

<hr /><em>This discussion was created from the release <a href='https://github.com/ZZ-Cat/CRSFforArduino/releases/tag/v1.0.0'>Version 1.0.0... at last!</a>.</em>

**Comment by SB-JHL-B26** [comment]
I need some help getting the Arduino IDE (2.3.2) to compile.

Compilation error: CRSFforArduino.hpp: No such file or directory.

Looks like the IDE is not seeing the CRSFforArduino library files.

Cheers, s

**Reply by ZZ-Cat** [reply]
That's odd.  
Can you elaborate on this a bit more, please?

**Reply by Kristabel28** [reply]
> Compilation error: CRSFforArduino.hpp: No such file or directory.


Hey! I think I know what the error is, but I'm not sure how to correct it. 

Normally for libraries you have library.h and library.cpp files, looks like an error in the extension here (*.hpp)

**Comment by SB-JHL-B26** [comment]
Hi, thank you for responding. I’m wondering if I have the Arduino IDE configured correctly. Here is what I have done so far … following the wiki …
I downloaded the 1.0.0 zip, extracted the zip, and then copied the “CRSFforArduino-1.0.0” folder to my Arduino/libraries folder. 

I opened a new sketch (File-->New Sketch) and added #include "CRSFforArduino.hpp", then compiled the new sketch which yielded "Compilation error: CRSFforArduino.hpp: No such file or directory."

Also, I do not see any examples from File-->Examples. It appears that the Arduino IDE does not see anything to do with CRSFdorArduino.


Cheers, Scott 

    On Friday, March 1, 2024 at 03:55:32 a.m. EST, ZZ-Cat ***@***.***> wrote:  
 
 


That's odd.
Can you elaborate on this a bit more, please?

**Comment by steinne** [comment]
I have had success by adding the library in th "sketch >> Include Library
>> Add .zip library  -
may be best to include this way...

[image: Screenshot 2024-03-01 at 12.20.39 PM.png]

**Reply by SB-JHL-B26** [reply]
Thank you for responding! How do I see the image you included?

**Reply by steinne** [reply]
Download the library as a ZIP
Open the Arduino IDE
Click Sketch
Click Include Library
Click Add .ZIP Library
Select the library to add
Navigate to and open the .ZIP file

**Reply by steinne** [reply]
<img width="782" alt="Screenshot 2024-03-01 at 12 20 39 PM" src="https://github.com/ZZ-Cat/CRSFforArduino/assets/8583803/20798045-3bc9-4ba9-a42d-17c9b5b7468b">

**Reply by SB-JHL-B26** [reply]
I downloaded the Source Code as a zip, added the zip but got a Error: 13 Library install failed ...

Perhaps I am not using the correct zip??

**Reply by steinne** [reply]
what is the error?     sometimes a line break is off in a properties file.       Send me the error, those are usually an easy fix

**Reply by SB-JHL-B26** [reply]
Error: 13 INTERNAL: Library install failed: moving extracted archive to destination dir: loading library.properties: error reading file: error parsing data at line 6: invalid line format, should be 'key=value'

**Reply by steinne** [reply]
that is an easy fix,   expand the zip,   and the open the library.properties file in a text editor.   The following line that starts with "paragraph"  has an extra line break in it.    Each key pair needs to be on a single line,   
Just make a small edit to update,  this line to only have a trailing line break,
/////
paragraph=This library enables you to connect either a TBS Crossfire or ExpressLRS receiver to your development board,giving you access to telemetry and up to 16 11-bit proportional RC channels over a tried-and-true serial protocol.
/////

 save the file.     Zip the folder and include as before.

**Reply by SB-JHL-B26** [reply]
Ok, thank you. I will try that a try. Thanks.

**Comment by SB-JHL-B26** [comment]
Update ...

Essentially, the Arduino IDE (2.3) is not seeing the CRSFforArduino library.

After unzipping the Source Code zip, I moving the top level folder to my C:\Users\WSB\Documents\Arduino\libraries folder.

I restarted IDE, went to Tools-->Manage Libraries to open the Library Manager. In the top search bar of the Library Manager, I typed in CRSF and three libraries were listed (Sam4uk, Jacob Williams and qqqlab).

There was no sign of a ZZ-Cat library to install. Perhaps this is because of the Caveat " ... This is because CRSF for Arduino is not yet available in the Arduino IDE's Library Manager.".

So I suppose the question is how do I install your libraries manually?

**Comment by SB-JHL-B26** [comment]
Update ... My experience with the CRSFforArduino code/libraries.

After trying a bunch of things I could not get the Adruino IDE (2.3) to see anything related to the CRSFforArduino libraries. Trying to run the examples manually resulted in "file not found" of course because the IDE did not have the necessary file paths established that comes with using the native IDE Library Manager. There is of course the possibility that I overlooked something simple. I am not a seasoned Arduino IDE user since in the past I have used Visual Studio. 

It would be beneficial to have this library use the default Arduino IDE Library installation tools.

For example, in the IDE Library Manager, typing "CRSF" yields 3 libraries. I simply clicked on the INSTALL button for AlfredoCRSF, and the IDE installed the library. Going to File-->Examples yielded 4 examples at the bottom "Examples from Custom Libraries" menu. 

I ran the "printAllChannels" example (after adjusting the Rx/Tx pin numbering for my Arduino Nano ESP32 board) and all worked perfectly with my RadioMaster RP2 ELRS Nano Receiver.

If there are some changes to the CRSFforArduino library installation process, I would be happy to test them out in order to contribute in some tiny way.

Thank you to those that took the time to help.

Cheers, Scott

**Comment by SB-JHL-B26** [comment]
Yes that worked! Thank you.

"rc_channels.ino" example compiles and loads but does not result in any output to the serial monitor.
I will need to do a bit more review and make sure the Rx/Tx pin designations are correct for my Arduino Nano ESP32 board.
Do you know where this configuration data resides?

Cheers, Scott 

    On Friday, March 1, 2024 at 12:47:46 p.m. EST, Neil Stein ***@***.***> wrote:  
 
 


that is an easy fix, expand the zip, and the open the library.properties file in a text editor. The following line that starts with "paragraph" has an extra line break in it. Each key pair needs to be on a single line,
Just make a small edit to update, this line to only have a trailing line break,
/////
paragraph=This library enables you to connect either a TBS Crossfire or ExpressLRS receiver to your development board,giving you access to telemetry and up to 16 11-bit proportional RC channels over a tried-and-true serial protocol.
/////

save the file. Zip the folder and include as before.

**Comment by ZZ-Cat** [comment]
## Compatibility issues with the Arduino IDE Library Manager

CRSF for Arduino is not yet available in the Library Manager in the Arduino IDE. I don't believe that this should be a reason why the Arduino IDE is not seeing CRSF for Arduino.  
It is also not yet available in PlatformIO's registry either. This is why I ask people to clone CRSF for Arduino directly off of its repository.

I have tested the compilation of CRSF for Arduino with the Arduino IDE version 1.8.10.  
I have since learned that this version of the Arduino IDE has been discontinued for quite some time.  

CRSF for Arduino version 1.0.0 is compatible with rev. 2.2 of the [Library Specifications.](https://arduino.github.io/arduino-cli/0.36/library-specification/)  
The documentation of the Library Specifications themselves aren't clear enough about how a library "should" look like, and the details are very vague or too generalised.  
Ergo, I had to "fill in the gaps" where the information in the specifications were lacking, by studying the layouts of other libraries (EG FastLED, and many of Adafruit's own libraries).

Rev. 2.2 of the Arduino Library Specifications is compatible with Arduino IDE version 1.8.10 and later.

## Possible confusion around installation and use with Arduino IDE

By rights, the folder structure of CRSF for Arduino should look like this:

```txt
|-libraries <- This is the Arduino IDE's folder where the rest of your libraries are stored.
    |
    |-CRSFforArduino <- Note the semantic version tag is omitted.
        |
        |-examples
        |    |-[code examples]
        |
        |-src
            |-[CRSF for Arduino's abstraction layers]
            |-CFA_Config.hpp
            |-CRSFforArduino.hpp <- This is what should be included in your sketches.
            |-CRSFforArduino.cpp <- This is what should be linked with CRSFforArduino.hpp when compiled
```
You may need to rename CRSF for Arduino's top level folder from `CRSFforArduino-1.0.0` to simply `CRSFforArduino`, once you have extracted the contents from the zip folder.

You may retain the rest of the files in CRSF for Arduino's directory.

You must not delete the `library.properties` file in your copy of CRSF for Arduino. If you delete this file, it will cause the Arduino IDE to default to "the old way" and look for files that don't exist. IE The Arduino IDE will assume that _all_ source and header files (`*.cpp` and `*.hpp`, respectively) are placed "flat" in CRSF for Arduino's folder... including CRSF for Arduino's abstraction layers. Thus, ignoring the `src` folder, nested sub-folders and nested files altogether; and the compilation process will fail.

---

## Typo in `library.properties`

I read up in the thread someone pointed out that I may have made a typo in the `library.properties` file, under the `paragraph` key. As stated, this is an easy fix.  
For me, I would remove all the line breaks in that key, and put what I wrote there all on one line.  
I may consolidate the `paragraph` at some point down the line.

## Library Manager

As previously stated, CRSF for Arduino is _not yet available_ in the Arduino IDE's Library Manager.  
Currently, I _am not_ aware of how to go about getting CRSF for Arduino into the Library Manager itself, nor am I aware of how Arduino developers manage that process.  
At some point down the line, I may look into how to go about getting CRSF for Arduino into the Library Manager.

After that, I may consider publishing CRSF for Arduino to PlatformIO's registry as well.

For the time being, you are advised to grab the latest version of the source code in the [Releases](https://github.com/ZZ-Cat/CRSFforArduino/releases) section.

## Further investigation is required

I am not entirely sure how this one got past all of my quality checks, let alone building and compiling locally in both Visual Studio Code and the Arduino IDE.

What I will do is, I will grab the latest version of the Arduino IDE; chuck CRSF for Arduino in there; test-compile and see what comes up. 
I will post screenshots of my results, and we can go from there.

## Additional

Recently, I migrated from Windows 11 to Pop!_OS. Therefore, I am required to set-up my own commit signing from scratch. This is where my current priority is at.  
I am doing this over in #87, and all the details for this process is in that Pull Request.  
Once I have finished that, I will look into this.

In the meantime, can @SB-JHL-B26 do the following:

1. Not use my email to contact me. 
  Sorry, I need to assert this.  
  My email should not be treated as my primary contact for issues regarding CRSF for Arduino.  
  I do not reply to emails. GitHub is the desired path, here.  
  My email is provided **for reference only.**  
  My primary point of contact regarding CRSF for Arduino is right here in its own repository via the Issues (for bug reports, feature requests, and development board compatibility requests) and Discussions (catch-all for what does not fit the Issues categories) tabs.
2. If this _is_ a bug (and _not_ a case of user-error), please consider heading over to my Issues tab, [creating a new Issue,](https://github.com/ZZ-Cat/CRSFforArduino/issues/new/choose) and clicking on "Submit a bug report".  
  This will put your issue into the development pipeline, and either myself or one of my contributors will get onto it when we are able to do so.

Thank you _very_ much for your engagement with us, so far.  
If you _do_ wish to contribute, you're very welcome to do so. But, before you do make your contributions, please take the time to familiarise yourself with my [Code of Conduct](https://github.com/ZZ-Cat/CRSFforArduino/blob/Main-Trunk/CODE_OF_CONDUCT.md) and [Contribution Guidelines.](https://github.com/ZZ-Cat/CRSFforArduino/blob/Main-Trunk/.github/CONTRIBUTING.md)

Keep in mind that Pull Requests with unverified and partially verified commits cannot be accepted, due to the fact that it is not possible to merge a Pull Request that contains one or more unverified or partially verified commits.  
All commits must be verified.  
For more information on commit signing, GitHub has [documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on it, plus instructions for how to set-up [GPG signing](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification) and [SSH signing](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#ssh-commit-signature-verification).  
SSH is simpler to set-up than GPG at the expense of the inability to revoke a key that is no longer in use (some may view this as a security vulnerability. You decide what best fits your needs).

**Reply by SB-JHL-B26** [reply]
A few points ...

Point 1) 

Removing the LF in the paragraph= line of library.properties as steinne suggested worked. After which the library was recognized. I am using Arduino IDE version 2.3.2. With the library.properties fixed, here are the steps that work and let the IDE Library Manager work correctly.

a) In the Documents/Arduino/libraries folder, create a folder called CRSFforArduino
b) Download the latest source code zip and place in a temporary location. Unzip the contents of the zip.
c) Manually copy the contents of the zip in the folder from step b) and paste them to the folder from step a).
d) Go to Sketch-->Include Library-->Contributed Libraries (at the bottom) and you will see CRSFforArduino.
e) Go to Files-->Examples-->Examples from Custom Libraries (at the bottom) and you will see CRSFforArduino. Expand that out CRSFforArduino and examples will be shown. Select and example and things will compile and load to your target.

Point 2)

After compiling and loading the example rc_channels sketch, there was no output. There is a detail regarding a setting in the Arduno IDE in Tools-->Pin Numbering--> "By Arduino pin (default)" and "By GPIO number (legacy)". I could not readily see anywhere in the code structure where it was obvious to verify which actual target Rx/Tx pins were being used. This is where my road ended today.

Point 3) 

I made the point earlier in this thread that I installed AlfredoCRSF library, with a simple tweak of the Rx/Tx pin numbers, ran an example and everything worked. I am not speaking to the actual quality/operation of the software but simply to the business of installing a custom library and running a provided example with very little fuss.

Summary ... My contribution to your CRSFforArduino project will be limited to high level testing. I would be more than happy to help in this regard so that when a release is done, it will "work right out of the box!".

Cheers, s

**Reply by ZZ-Cat** [reply]
Interesting. :thinking: 

I have _not yet_ touched Arduino IDE 2.x.x. This will change as _soon_ as I have it installed on my end.  
Clearly a _lot_ has changed since 1.8.10, and I need to play "catch up" with this.

CRSF for Arduino has a constructor `CRSFforArduino::CRSFforArduino(HardwareSerial *serialPort)` where you can provide your own pre-defined and pre-configured UART to CRSF for Arduino. This includes your own pins... in most cases.

ESP32 is completely different to other platforms.  
In the context of UART, it looks as though the pins are defined when you call `begin()` on the platform.  
Whereas with SAMD21 and SAMD51, you define the pins in the `HardwareSerial` constructor and `begin()` is used to simply call the baudrate and configuration (EG 420000 and 8-bits, no parity, and 1 stop bit).  
The last person that requested ESP32 support did not hang around to help me test the codebase on their targets to ensure everything was working. I foolishly made the assumption that "it just works", and merged it.... only to come up with what you're experiencing now.

One of the primary goals of this project _is_ to get it to a state where it works out-of-the-box, and you should not have to unnecessarily fuck with it just to get it working.  
This is where I need folks what can do hardware testing on supported development boards come in. 'Cause the available targets that CRSF for Arduino is now compatible with... that list has grown far beyond any of the limited hardware I have available on hand, and it isn't realistic for me to test _all_ the development boards myself.

**Comment by SB-JHL-B26** [comment]
Update ... With the help from 'steinne', I was able to install the CRSFforArduino library, compile and load. In particular the "rc_channels" example.

Unfortunately my Arduino Nano ESP32 board does not output anything. "CRSF for Arduino initialisation failed!" or "RC Channels Example" or "Ready" did not print in the Serial Monitor and there is no channel data being displayed.

I have 2 sketches open on and can switch back and forth between "AlfredoCRSF" and "CRSFforArduio" examples. Same hardware same IDE. The AlfredoCRSF example sketch provided channel data and the CRSFforArduino example sketch does not print anything.

At this point I have spent quite a bit time testing and trying to get CRSFforArduino to work. I will have to leave it for now since I believe some things require attention which are beyond my "user" level capabilities. 

I will continue to test when the code/installation base is more mature and a couple of issues are further tested by the devs. 

I will commit my time to testing as a contribution to this cause. 

Cheers, s

**Comment by SB-JHL-B26** [comment]
Thank you for responding. I will take a look at HardwareSerial with the aim of getting it to work. I will post my results.

It is quite a task to preside over such a large number of boards and environments! 

I will help test the Arduino Nano ESP32-S3 board as well as test the Arduino IDE. Just give me a heads up when new releases happen. 

Cheers, Scott 

    On Friday, March 1, 2024 at 05:27:11 p.m. EST, ZZ-Cat ***@***.***> wrote:  
 
 


Interesting. 🤔

I have not yet touched Arduino IDE 2.x.x. This will change as soon as I have it installed on my end.
Clearly a lot has changed since 1.8.10, and I need to play "catch up" with this.

CRSF for Arduino has a constructor CRSFforArduino::CRSFforArduino(HardwareSerial *serialPort) where you can provide your own pre-defined and pre-configured UART to CRSF for Arduino. This includes your own pins... in most cases.

ESP32 is completely different to other platforms.
In the context of UART, it looks as though the pins are defined when you call begin() on the platform.
Whereas with SAMD21 and SAMD51, you define the pins in the HardwareSerial constructor and begin() is used to simply call the baudrate and configuration (EG 420000 and 8-bits, no parity, and 1 stop bit).
The last person that requested ESP32 support did not hang around to help me test the codebase on their targets to ensure everything was working. I foolishly made the assumption that "it just works", and merged it.... only to come up with what you're experiencing now.

One of the primary goals of this project is to get it to a state where it works out-of-the-box, and you should not have to unnecessarily fuck with it just to get it working.
This is where I need folks what can do hardware testing on supported development boards come in. 'Cause the available targets that CRSF for Arduino is now compatible with... that list has grown far beyond any of the limited hardware I have available on hand, and it isn't realistic for me to test all the development boards myself.

**Reply by ZZ-Cat** [reply]
Awesome!

Generally, when someone lets me know they are interested in helping out with testing hardware, I tag them in a related Pull Request and I provide instructions for them to check out the topic branch and flash their target(s) with the firmware, and they provide me with feedback on how it all goes; and I make adjustments to the codebase, along the way.

**Comment by SB-JHL-B26** [comment]
I have a few boards that I can help out with. I will elaborate tomorrow. Cheers, Scott.On Mar 1, 2024, at 6:40 PM, ZZ-Cat ***@***.***> wrote:﻿
Awesome!
Generally, when someone lets me know they are interested in helping out with testing hardware, I tag them in a related Pull Request and I provide instructions for them to check out the topic branch and flash their target(s) with the firmware, and they provide me with feedback on how it all goes; and I make adjustments to the codebase, along the way.

**Comment by ZZ-Cat** [comment]
@SB-JHL-B26. While I was working on #91, I noticed that my [Arduino build check](../actions/workflows/arduinoide_build.yml) was [failing](../actions/runs/8137336166/job/22235431691) with the same error you mentioned here.

Due to the severity of this bug, I have issued a [PSA](../discussions/92) and created an [Issue](../issues/93) from this. Thus putting it at the top of my list in the development pipeline. A fix for this will be available in the next patch release.

This got past my quality control, because at the time, there was a bug in the Arduino build check workflow on Adafruit's end and I had to temporarily disable it until they got it fixed.  
Since they fixed it, I didn't reinstate the build check on my end. So, this bug slipped through my quality control unnoticed.  
Lesson learned: _Don't_ disable quality control checks! Wait for the maintainers of those checks to release fixes for any bugs that crop up on their end, and _don't_ rush development of my own project to meet my own perceived "dead-lines".

I have since reinstated that build check, and it picked the error you mentioned up straight away.  
Hence why I have made this my top priority to get it sorted.

**Reply by SB-JHL-B26** [reply]
Stuff happens and it is OK! It sounds like you have a good handle on this. As soon as I see the patch release, I will test it and post my findings. Cheers, s

**Reply by ZZ-Cat** [reply]
> Stuff happens and it is OK! It sounds like you have a good handle on this. As soon as I see the patch release, I will test it and post my findings. Cheers, s 

Excellent. =^/.^=

I am still having issues with my Arduino CI workflow. However, I was able to successfully get this [branch](../tree/broken-arduino-ide-build) of CRSF for Arduino to build in the Arduino IDE v2.3.2 on your target with impunity.

I am making a cut of Version 1.0.1 today, and it will be available in the [Releases](../releases) section shortly.
